// feathers-telegram-bot/lib/hooks/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _authenticate = require('./authenticate');

var _authenticate2 = _interopRequireDefault(_authenticate);

var _restrictToTelegram = require('./restrict-to-telegram');

var _restrictToTelegram2 = _interopRequireDefault(_restrictToTelegram);

var _restrictToChatAdmin = require('./restrict-to-chat-admin');

var _restrictToChatAdmin2 = _interopRequireDefault(_restrictToChatAdmin);

var _restrictToChatMember = require('./restrict-to-chat-member');

var _restrictToChatMember2 = _interopRequireDefault(_restrictToChatMember);

var _patchOrCreateMessageUsers = require('./patch-or-create-message-users');

var _patchOrCreateMessageUsers2 = _interopRequireDefault(_patchOrCreateMessageUsers);

var _patchMessageUserChat = require('./patch-message-user-chat');

var _patchMessageUserChat2 = _interopRequireDefault(_patchMessageUserChat);

var _patchOrCreateMessageChats = require('./patch-or-create-message-chats');

var _patchOrCreateMessageChats2 = _interopRequireDefault(_patchOrCreateMessageChats);

var _isTelegram = require('./is-telegram');

var _isTelegram2 = _interopRequireDefault(_isTelegram);

var _isBotCommand = require('./is-bot-command');

var _isBotCommand2 = _interopRequireDefault(_isBotCommand);

var _isChatType = require('./is-chat-type');

var _isChatType2 = _interopRequireDefault(_isChatType);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  authenticate: _authenticate2.default,
  restrictToTelegram: _restrictToTelegram2.default,
  restrictToChatAdmin: _restrictToChatAdmin2.default,
  restrictToChatMember: _restrictToChatMember2.default,
  patchOrCreateMessageUsers: _patchOrCreateMessageUsers2.default,
  patchMessageUserChat: _patchMessageUserChat2.default,
  patchOrCreateMessageChats: _patchOrCreateMessageChats2.default,
  isTelegram: _isTelegram2.default,
  isBotCommand: _isBotCommand2.default,
  isChatType: _isChatType2.default
};
module.exports = exports['default'];
