// feathers-telegram-bot/lib/hooks/patch-message-user-chat.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = patchMessageUserChat;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function patchMessageUserChat() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  return function (hook) {
    options = Object.assign({
      telegramIdField: 'telegramId',
      service: 'users',
      as: 'telegram.chats'
    }, options);
    var service = hook.app.service(options.service);
    if (hook.params.telegram) {
      var id = hook.params.message.from.id;
      var chat = hook.params.message.chat;
      return service.find({
        query: _defineProperty({}, options.telegramIdField, id),
        paginate: false
      }).then(function (data) {
        if (data.length) {
          var user = data[0];
          var currentChats = user[options.as] || {};
          if (!user[options.as] || !user[options.as][chat.id]) {
            return service.patch(user[service.id], _defineProperty({}, options.as + '.' + chat.id, Object.assign(currentChats[chat.id] || {}, {}))).then(function () {
              return hook;
            });
          }
        }
        return hook;
      });
    }
    return hook;
  };
}
module.exports = exports['default'];
