// feathers-telegram-bot/lib/hooks/patch-or-create-message-chats.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = patchOrCreateMessageChats;

var _lodash = require('lodash.omit');

var _lodash2 = _interopRequireDefault(_lodash);

var _lodash3 = require('lodash.union');

var _lodash4 = _interopRequireDefault(_lodash3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function patchOrCreateMessageChats() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  return function (hook) {
    options = Object.assign({
      service: 'chats',
      userAs: 'users',
      all: true
    }, options);
    var service = hook.app.service(options.service);
    var patchOrCreate = function patchOrCreate(chat) {
      return service.find({
        query: _defineProperty({}, service.id, chat.id),
        paginate: false
      }).then(function (data) {
        if (data.length) {
          if (chat[options.userAs]) {
            chat[options.userAs] = (0, _lodash4.default)(data[0][options.userAs] || [], chat[options.userAs]);
          }
          var props = Object.keys(chat);
          var changed = props.some(function (prop) {
            return JSON.stringify(data[0][prop]) != JSON.stringify(chat[prop]);
          });
          if (changed) {
            return service.patch(data[0][service.id], (0, _lodash2.default)(chat, ['id']));
          } else {
            return Promise.resolve(data[0]);
          }
        } else {
          return service.create(Object.assign(_defineProperty({}, service.id, chat.id), (0, _lodash2.default)(chat, ['id'])));
        }
      });
    };
    if (hook.params.telegram) {
      var promises = [];
      var message = hook.params.message;
      if (message.chat) {
        var chat = Object.assign({}, message.chat);
        if (options.userAs) {
          chat[options.userAs] = [message.from.id];
        }
        promises.push(patchOrCreate(chat));
      }
      if (options.all && message.forward_from_chat) promises.push(patchOrCreate(message.forward_from_chat));
      return Promise.all(promises).then(function () {
        return hook;
      });
    }
    return hook;
  };
}
module.exports = exports['default'];
